docker/server/build.gradle

```diff
@@ -0,0 +1,92 @@
+import io.deephaven.tools.License
+
+plugins {
+    id 'com.bmuschko.docker-remote-api'
+}
+
+def baseMap = [
+    'server-base': 'server',
+    'nltk-base': 'server-nltk',
+    'pytorch-base': 'server-pytorch',
+    'sklearn-base': 'server-sklearn',
+    'tensorflow-base': 'server-tensorflow',
+]
+
+baseMap.keySet().each {base ->
+    evaluationDependsOn Docker.registryProject(base)
+}
+
+evaluationDependsOn ':deephaven-jpy'
+evaluationDependsOn ':deephaven-wheel'
+evaluationDependsOn ':deephaven2-wheel'
+
+configurations {
+    serverApplicationDist
+}
+
+dependencies {
+    serverApplicationDist project(path: ':grpc-api-server-native', configuration: 'applicationDist')
+}
+
+def dockerLicenses = License.createFrom(project).syncDockerLicense().get()
+
+def context = { String base ->
+    project.layout.buildDirectory.dir("${baseMap[base]}/context")
+}
+
+def prepareDockerClosure = { String base ->
+    project.tasks.register("prepareDocker-${baseMap[base]}", Sync) {
+        // TODO(deephaven-core#1596): Remove extra dependencies for build-ci.yml
+        // Currently, GH build-ci.yml calls dockerCreateDockerfile, and relies on buildx; so we need to
+        // make sure that dockerCreateDockerfile has the proper dependencies.
+        // If we change how our build works in the future (so that gradle is responsible for pushing the
+        // image), we can remove this extra dependency.
+        dependsOn Docker.registryTask(project, base)
+
+        // deephaven-jpy.whl
+        def deephavenJpyWheel = project(':deephaven-jpy').tasks.getByName('buildWheel')
+
+        // deephaven.whl
+        def deephavenWheel = project(':deephaven-wheel').tasks.getByName('buildWheel')
+
+        // deephaven2.whl
+        def deephaven2Wheel = project(':deephaven2-wheel').tasks.getByName('buildWheel')
+
+        from 'src/main/common'
+        from "src/main/${baseMap[base]}"
```

Comment:
this looks like the only different 'from', everything else coming into this sync is the same, right? could we take advantage of that somehow, and not end up with both so many copies of these files (the applicationDist for example isn't small) and also dependencies such that any chance to any of them require syncing again (i.e. push them down into the DockerBuildImage task dependencies?

I can't obviously see a way to do this with very small changes, but consider this - just one sync task, with this rough structure
```
docker/
  common/
    *.whl
    serverApplicationDist/server-1.2.3.tar
    licenses/
    ... (unroll common here too)
  server/
  server-nltk/
  server-pytorch/
  (etc)
```
Then in the dockerfile, since we already have $BASE, can we point at the one directory as the docker context for each build? granted, we still have to upload each time to docker as the context, but we had to do that already anyway.

Reply:
Good idea. I think this will be easy.